src/controller/Controller.cpp

```diff
@@ -19,6 +19,79 @@ Controller::Controller() : mIsRumbling(false) {
     }
 }
 
+int8_t Controller::ReadStick(int32_t virtualSlot, Stick stick, Axis axis) {
+    switch (stick) {
+        case Stick::LEFT: {
+            switch (axis) {
+                case Axis::X: {
+                    if (getLeftStickX(virtualSlot) == 0) {
+                        if (getPressedButtons(virtualSlot) & BTN_STICKLEFT) {
+                            return -MAX_AXIS_RANGE;
+                        } else if (getPressedButtons(virtualSlot) & BTN_STICKRIGHT) {
+                            return MAX_AXIS_RANGE;
+                        }
+                    } else {
+                        return getLeftStickX(virtualSlot);
+                    }
+                }
+                case Axis::Y: {
+                    if (getLeftStickY(virtualSlot) == 0) {
+                        if (getPressedButtons(virtualSlot) & BTN_STICKDOWN) {
+                            return -MAX_AXIS_RANGE;
+                        } else if (getPressedButtons(virtualSlot) & BTN_STICKUP) {
+                            return MAX_AXIS_RANGE;
+                        }
+                    } else {
+                        return getLeftStickY(virtualSlot);
+                    }
+                }
+            }
+        }
+        case Stick::RIGHT: {
+            switch (axis) {
+                case Axis::X: {
+                    if (getRightStickX(virtualSlot) == 0) {
+                        if (getPressedButtons(virtualSlot) & BTN_VSTICKLEFT) {
+                            return -MAX_AXIS_RANGE;
+                        } else if (getPressedButtons(virtualSlot) & BTN_VSTICKRIGHT) {
+                            return MAX_AXIS_RANGE;
+                        }
+                    } else {
+                        return getRightStickX(virtualSlot);
+                    }
+                }
+                case Axis::Y: {
+                    if (getRightStickY(virtualSlot) == 0) {
+                        if (getPressedButtons(virtualSlot) & BTN_VSTICKDOWN) {
+                            return -MAX_AXIS_RANGE;
+                        } else if (getPressedButtons(virtualSlot) & BTN_VSTICKUP) {
+                            return MAX_AXIS_RANGE;
+                        }
+                    } else {
+                        return getRightStickY(virtualSlot);
+                    }
+                }
+            }
+        }
+    }
+
+    return 0;
+}
+
+void Controller::ProcessStick(int8_t& x, int8_t& y, uint16_t deadzone) {
+    // create scaled circular dead-zone in range {-15 ... +15}
+    auto len = sqrt(x * x + y * y);
+    if (len < deadzone) {
+        len = 0;
+    } else if (len > MAX_AXIS_RANGE) {
+        len = MAX_AXIS_RANGE / len;
+    } else {
+        len = (len - deadzone) * MAX_AXIS_RANGE / (MAX_AXIS_RANGE - deadzone) / len;
+    }
+    x *= len;
+    y *= len;
+}
```

Comment:
What happened to the octagonal gate processing that should probably be in this function?

Reply:
Can you be a little more specific? It seems like the octagonal gating code was different to this code that scales and deadzones.